clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -951,28 +959,130 @@ Sema::ActOnDecompositionDeclarator(Scope *S, Declarator &D,
   return New;
 }
 
+namespace {
+// CheckBindingsCount
+//  - Checks the arity of the structured bindings
+//  - Creates the resolved pack expr if there is
+//    one
+bool CheckBindingsCount(Sema &S, DecompositionDecl *DD, QualType DecompType,
+                        ArrayRef<BindingDecl *> Bindings,
+                        unsigned MemberCount) {
+  auto BindingWithPackItr =
+      std::find_if(Bindings.begin(), Bindings.end(),
+                   [](BindingDecl *D) -> bool { return D->isParameterPack(); });
+  bool HasPack = BindingWithPackItr != Bindings.end();
+  bool IsValid;
+  if (!HasPack) {
+    IsValid = Bindings.size() == MemberCount;
+  } else {
+    // there may not be more members than non-pack bindings
+    IsValid = MemberCount >= Bindings.size() - 1;
+  }
+
+  if (IsValid && HasPack) {
+    TemplateTypeParmDecl *DummyTemplateParam = TemplateTypeParmDecl::Create(
+        S.Context, S.Context.getTranslationUnitDecl(),
+        /*KeyLoc*/ SourceLocation(), /*NameLoc*/ SourceLocation(),
+        /*TemplateDepth*/ 0, /*AutoParameterPosition*/ 0,
+        /*Identifier*/ nullptr, false, /*IsParameterPack*/ true);
```

Comment:
Can you explain the logic here?

Reply:
We are sort of creating an implicit template and need a dummy template parameter. It is really only used to create the pack expansion type. This is to create an Expr that the BindingDecl is "aliasing".

Reply:
Did you try to use a dummy type like `DependentTy` ?
(ie it is not immediately clear to me that the pattern of an expansion type needs to be a template parameter)

Reply:
The first line of `getPackExpansionType` has an assert on `Pattern->containsUnexpandedParameterPack()`. Now that I am looking at it, there is a `ExpectsPackInType` parameter. Maybe that was not there before or I just missed it. I will try that.